Support new cuFFT callbacks

leofang opened the pull request:
cuFFT now supports a new callback mechanism that is not only more performant but also more friendly to dynamic languages like Python and Julia, currently under [public preview](https://docs.nvidia.com/cuda/cufft/ltoea/index.html) and will be released in a future CUDA version. 

The new approach does not require any ugly compilation of host & device code; instead, we just need to compile a user-provided device function (currently supporting raw CUDA C++ strings; once we support user-provided types in `@cupyx.jit.rawkernel` (https://github.com/cupy/cupy/issues/6663) we can also support pure Python functions as device functions) into LTO IR, and pass it to the new cuFFT API `cufftXtSetJITCallback` which will be responsible for linking the LTO IR with the cuFFT kernel.

The existing callback manager (`_CallbackManager`) is renamed to `_LegacyCallbackManager`, to be distinguished with the new `_JITCallbackManager`. My suggestion is once this feature is officially released (generally accessible), we add a deprecation warning to `_LegacyCallbackManager`. Users can currently pick either callback approach via the new option `cp.fft.config.set_cufft_callbacks(..., cb_ver=...)`.

Currently, this feature requires users loading libcufft EA version: `LD_PRELOAD=/path/to/EA/libcufft.so python ...`, including running the test suite or sample code, so that the libcufft copy from CUDA Toolkit is not used.

jakirkham commented:
Stumbled across this recently when doing some testing with CuPy. This seems like a nice addition

Also it looks like this is now out of preview. Meaning we could start relying on this

> The cuFFT LTO EA preview is deprecated and will be removed in the near future.
> 
> The main feature of the preview, LTO callbacks, has been integrated into cuFFT as part of the [CUDA Toolkit 12.6 Update 2](https://docs.nvidia.com/cuda/cufft/#cufft-callback-routines). This update comes with significant improvements compared to the cuFFT LTO EA preview, including the ability of giving callback device functions arbitrary names, as well as improved performance.

ref: https://docs.nvidia.com/cuda/cufft/ltoea/index.html

leofang commented:
/test mini

leofang commented on .pfnci/matrix.yaml:
```diff
@@ -730,26 +730,26 @@
-- project: "cupy.linux.cuda11x-cuda-python"
-  target: "cuda11x-cuda-python"
+# CUDA 12.x / CUDA Python | Linux
+- project: "cupy.linux.cuda12x-cuda-python"
```
@kmaehashi I need to move the cuda-python CI to 12.x because we need a new NVRTC API from 12.x. But it seems a new `cupy.linux.cuda12x-cuda-python` pipeline cannot be registered by making changes in the codebase? I assume there's a backend that need to be updated as well? 🙂

leofang replied:
Thanks, Kenichi-san! I'll rerun the CI after 13.6.0 is out, to avoid resource contention.

leofang commented:
/test mini

leofang commented:
All tests pass! The new callback supports both Linux/Windows, and I've adjusted the codebase to make it tested on both platforms. Oh, it also supports T4 so we don't have to skip the tests in the CI 😉 

@kmaehashi this is ready for review (modulo https://github.com/cupy/cupy/pull/8242#discussion_r2270019578). I've added deprecation warnings to the old/legacy callback path because the new one is much more accessibly (see above) and performant.

leofang commented:
@kmaehashi I wonder why `cupy.linux.cuda11x-cuda-python` still shows up in the CI list, but not the cuda12x one?

leofang commented:
Friendly nudge, @kmaehashi 🙂

kmaehashi commented:
Fixed in https://github.com/cupy/cupy/pull/9368, new tests added in PRs can now be trigged within that PR.

leofang commented:
This looks like OOM to me, not sure if it's flaky on Windows @kmaehashi?
```
=========================== short test summary info ===========================
FAILED cupy_tests\core_tests\test_raw.py::TestRawPicklable_param_3_{compile=True, raw='ker'}::test_raw_picklable - subprocess.CalledProcessError: Command '['C:\\Development\\Python\\Python310\\python.exe', 'TestRawPicklable.py']' returned non-zero exit status 3221226505.
FAILED cupy_tests\core_tests\test_raw.py::TestRawPicklable_param_4_{compile=True, raw='mod'}::test_raw_picklable - subprocess.CalledProcessError: Command '['C:\\Development\\Python\\Python310\\python.exe', 'TestRawPicklable.py', 'test_sum']' returned non-zero exit status 3221226505.
FAILED example_tests\test_custom_struct.py::TestCustomStruct::test_builtin_vectors - subprocess.CalledProcessError: Command '('C:\\Development\\Python\\Python310\\python.exe', 'C:\\Windows\\TEMP\\flexci\\run-00410212\\work\\src\\tests\\example_tests\\..\\..\\examples\\custom_struct/builtin_vectors.py')' returned non-zero exit status 3221226505.
= 3 failed, 104202 passed, 8043 skipped, 768 deselected, 776 xfailed, 2582 warnings in 3502.94s (0:58:22) =
```

leofang commented:
ping @kmaehashi 😅

leofang commented:
/test win

leofang commented:
/test mini

leofang commented:
CI is green! 🎉

leofang commented:
ping @asi1024

asi1024 commented on cupy/fft/_callback.pyx:
```diff
@@ -467,53 +498,215 @@ cdef class _CallbackManager:
+
+# LHS is our legacy convention
+# RHS is a valid C identifier pattern
+cdef object cupy_callback_pattern = re.compile(
+    r'd_(?:load|store)CallbackPtr\s*=\s*([A-Za-z_]\w*)\s*;')
```
How about separating the patterns to handle cases where users might pass arguments in the wrong order?

```suggestion
cdef object cupy_load_callback_pattern = re.compile(
    r'd_loadCallbackPtr\s*=\s*([A-Za-z_]\w*)\s*;')

cdef object cupy_store_callback_pattern = re.compile(
    r'd_storeCallbackPtr\s*=\s*([A-Za-z_]\w*)\s*;')
```

leofang commented:
@asi1024 ready for re-review!

leofang commented:
Looks like @kmaehashi also needs to approve 🙂
